packages/admin/admin/src/router/PromptHandler.tsx

```diff
@@ -5,13 +5,94 @@ import { Prompt } from "react-router";
 import { PromptAction, RouterConfirmationDialog } from "./ConfirmationDialog";
 import { RouterContext } from "./Context";
 
-interface IMessages {
-    [id: string]: (location: History.Location, action: History.Action) => boolean | string;
+interface PromptHandlerState {
+    showConfirmationDialog: boolean;
+    message: string;
+    callback?: (ok: boolean) => void;
+}
+export interface PromptHandlerApi {
+    showDialog: (message: string, callback: (ok: boolean) => void) => void;
+}
+function InnerPromptHandler({
+    registeredMessages,
+    saveActions,
+    apiRef,
+}: {
+    registeredMessages: React.MutableRefObject<PromptMessages>;
+    saveActions: React.MutableRefObject<SaveActions>;
+    apiRef: React.MutableRefObject<PromptHandlerApi | undefined>;
+}) {
+    const [state, setState] = React.useState<PromptHandlerState>({
+        showConfirmationDialog: false,
+        message: "",
+        callback: undefined,
+    });
+    if (apiRef)
+        apiRef.current = {
+            showDialog: (message: string, callback: (ok: boolean) => void) => {
+                setState({
+                    showConfirmationDialog: true,
+                    message,
+                    callback,
+                });
+            },
+        };
+
+    const promptMessage = (location: History.Location, action: History.Action): boolean | string => {
+        let ret: boolean | string = true;
+        Object.keys(registeredMessages.current).forEach((id) => {
+            if (location.pathname.startsWith(registeredMessages.current[id].path)) {
+                // allow transition if location is below path where prompt was rendered
+                return true;
```

Comment:
This only returns in `Object.keys(registeredMessages.current).forEach(...)`, is this intentional? If not, you'll need to use `for .. of`

Reply:
hm, yes.... but how is this working correclty??

Reply:
because `ret` in the outer scope is true by default, and you only change it below this return. The return in the forEach isn't used for anything